synapse/storage/state.py

```diff
@@ -259,10 +263,19 @@ def _get_state_groups_from_groups_txn(self, txn, groups, types=None):
                         key = (typ, state_key)
                         results[group][key] = event_id
         else:
+            where_args = []
             if types is not None:
-                where_clause = "AND (%s)" % (
-                    " OR ".join(["(type = ? AND state_key = ?)"] * len(types)),
-                )
+                where_clause = "AND ("
+                for typ in types:
+                    if typ[1] is None:
+                        where_clause += "(type = ?)"
+                        where_args.extend(typ[0])
+                    else:
+                        where_clause += "(type = ? AND state_key = ?)"
+                        where_args.extend([typ[0], typ[1]])
+                    if typ != types[-1]:
```

Comment:
is this supposed to be checking if it's the last element? it looks fragile.

Why not do `" OR ".join(where_clauses)`, or something

Reply:
yes, it is. guess it is fragile if you have dups in the list. stupid brainfart on .join(where_clauses)